src/cast/safe/mod.rs

```diff
@@ -0,0 +1,56 @@
+//! Traits for safe casting in a generic context.
+//!
+//! Each of the traits herein has a varying degree of safety. Raw casting is
+//! implicitly dangerous because it can perform four possible actions:
+//! * Convert from integer to float or vice versa.
+//! * Convert from signed to unsigned or vice versa.
+//! * Increases the size (or precision) of a number without loss.
+//! * Decreases the size (or precision) of a number with loss.
+//!
+//! By using the raw cast operation you may get unintentional side effects.
+//! These will often compile without warning and appear only as bugs later.
+//!
+//! The purpose of this module is to provide zero-cost abstraction traits that
+//! encapsulate each discrete action of a cast. When using the methods on these
+//! traits, you will only ever get the intended explicit behavior.
+//!
+//! For example, here is a generic implementation of a trivial function which
+//! takes any integer type that can hold at least 12 bits and returns the 12
+//! least significant bits. Note that attempting to pass an 8-bit integer to
+//! this function would result in a compiler error (since `TrimInto<u16>` is
+//! not implemented for `u8`).
+//!
+//! ```rust
+//! use num_traits::cast::safe::{SignCast, TrimInto};
+//!
+//! const MASK: u16 = 0b0000111111111111;
+//!
+//! fn low12<T: TrimInto<u16>, C: SignCast<Unsigned=T>>(x: C) -> u16 {
+//!     x.unsigned().trim() & MASK
+//! }
```

Comment:
Since this is your only concrete usage example, how is it much better than:

```rust
fn low12<T: AsPrimitive<u16>, C: SignCast<Unsigned=T>>(x: C) -> u16 {
    x.unsigned().as_() & MASK
}
```

Or personally, I'd write the bounds like this:

```rust
fn low12<T: SignCast>(x: T) -> u16
where
    T::Unsigned: AsPrimitive<u16>,
{
    x.unsigned().as_() & MASK
}
```

This doesn't exclude `u8` input, but is that really necessary?

Reply:
Yes. Think of a nuclear silo. Pretend there is a flag that controls some aspect of launching the rocket in bit 11. Do you want people passing a small integer into the function and getting unexpected behavior?

I'd rather the interface clearly define: 16 bits are the minimum. The compiler will fail otherwise.